### PR TITLE
Add specific source types

### LINKED ISSUE #17: Create source types for each type of Heroku log drain message
For a Heroku log drain there are at least four different types of messages. See https://devcenter.heroku.com/articles/logging#runtime-logs.

1. App logs (i.e. the console.log calls)
2. System logs (e.g. Heroku router logs)
3. API logs (e.g. dyno scaling events)
4. [Heroku runtime metrics](https://devcenter.heroku.com/articles/log-runtime-metrics)
5. Add-on logs

Each type of log should be assigned their own source type.

It may be possible to split the app logs into two source types, one for JSON based messages and another for field-value messages.

Add as a source type transform below the following:

https://github.com/Financial-Times/splunk-heroku/blob/1ab4b0c2664ba66a90eebd77f9f7dd8e1d16e496/app/default/props.conf#L13

### PR DESCRIPTION
**Pull request by sjparkinson**
Resolves #17.

Enabled field extraction for JSON based messages and adds additional source types for the various Heroku platform logs.

~I have a call with Splunk on Friday morning and want to run this past them.~

~This doesn't resolve the automatic JSON field extraction.~

**Screenshots**

Screenshot 2022-06-10 at 17 40 40

Screenshot 2022-06-10 at 17 43 00

### CODE REVIEW

File: app/default/props.conf [hunk 1]
```diff
@@ -10,12 +10,46 @@ TIME_FORMAT = %Y-%m-%dT%T.%6N%:z
 # Strip the RFC5424 header from the _raw field.
 SEDCMD-rfc5424_header = s/^\d+\s<\d{1,3}>\d\s(.+)\shost\s/\1 /
 
-TRANSFORMS-null = heroku_requests_doubleunderscore_null
+TRANSFORMS-source_type = heroku_system_source_type, heroku_app_source_type, heroku_requests_source_type, heroku_api_source_type, heroku_app_json_source_type
```
Comment (sjparkinson) [inline app/default/props.conf:13]: The Splunk documentation suggests this shouldn't work in Splunk Cloud, but I've been told by Lars that it should. Other Splunk apps make use of this pattern.

File: app/default/props.conf [hunk 2]
```diff
@@ -10,12 +10,46 @@ TIME_FORMAT = %Y-%m-%dT%T.%6N%:z
...
 EXTRACT-heroku_log_format = ^\S+\s(?<component>\S+)\s(?<dyno>\S+)
 
 # For JSON strip header values to enable syntax highlighting.
 EVAL-_raw = replace(_raw, "^.+\s-\s(\{.+\})\s*$", "\1")
 
-# Look up the Heroku error code description, see https://devcenter.heroku.com/articles/error-codes.
-LOOKUP-heroku_error = heroku_error_lookup code OUTPUTNEW desc AS code_desc
+# Based on the `_json` source type.
+INDEXED_EXTRACTIONS = json
+KV_MODE = none
```
Comment (sjparkinson) [inline app/default/props.conf:55]: These don't seem to do anything.

I've also tried `KV_MODE = json`, no luck with that either.

File: app/default/props.conf [hunk 3]
```diff
@@ -10,12 +10,46 @@ TIME_FORMAT = %Y-%m-%dT%T.%6N%:z
...
+EXTRACT-heroku_log_format = ^\S+\s(?<component>\S+)\s(?<dyno>\S+)
+
+[heroku_app_json]
+category = Structured
+description = JSON based logging output from your Heroku app, including logs generated by your code and dependencies.
 
 EXTRACT-heroku_log_format = ^\S+\s(?<component>\S+)\s(?<dyno>\S+)
 
 # For JSON strip header values to enable syntax highlighting.
 EVAL-_raw = replace(_raw, "^.+\s-\s(\{.+\})\s*$", "\1")
```
Comment (sjparkinson) [inline app/default/props.conf:51]: I've tried using `SEDCMD` instead with the same pattern, but that doesn't seem to do anything, less that EVAL which at least gets the formatting working well.

File: app/default/transforms.conf [hunk 4]
```diff
@@ -1,9 +1,35 @@
...
+[heroku_requests_null_queue]
 REGEX = router\s-\s.+path="\/__(gtg|health)"\s.+status=200
 DEST_KEY = queue
 FORMAT = nullQueue
 
+# Various types of log messages for Heroku log drains. See https://devcenter.heroku.com/articles/logging#runtime-logs for more information.
+[heroku_app_source_type]
+REGEX = ^\S+\sapp\s
+FORMAT = sourcetype::heroku_app
+DEST_KEY = MetaData:Sourcetype
```
Comment (sjparkinson) [inline app/default/transforms.conf:11]: Apparently this is required?

```suggestion
DEST_KEY = MetaData:Sourcetype
WRITE_META = true
```

Comment (sjparkinson) [reply app/default/transforms.conf:11]: But locally it all seems to worth without this 🤷

File: app/default/props.conf (hunk 3 above)
Comment (sjparkinson) [reply app/default/props.conf]: Got `SEDCMD` working, but needed to run on the `heroku` source type.

File: app/default/props.conf [hunk 5]
```diff
@@ -1,21 +1,42 @@
...
 
 # Look up the Heroku error code description, see https://devcenter.heroku.com/articles/error-codes.
-LOOKUP-heroku_error = heroku_error_lookup code OUTPUTNEW desc AS code_desc
+LOOKUP-heroku_code_desc = heroku_error_lookup code OUTPUTNEW desc AS code_desc
+
+# Note: Only extract component as request logs contain a dyno field with a value that aligns with the other Heroku source types. Would always be "router" otherwise.
+EXTRACT-heroku_component = ^\S+\s(?<component>\S+)
+
+# Do not index some high volume low value messages.
+TRANSFORMS-heroku_requests_null_queue = heroku_requests_null_queue
```
Comment (sjparkinson) [inline app/default/props.conf:36]: Makes me think we should do something similar for logs coming from Fastly!

**Comment by sjparkinson** [comment]
@jonnangle mind taking on the review for this? Happy to run through any bits on a call too

**Review by jonnangle — APPROVED** [review]

**Pull request merged**